Update `tar`, `flate2`, and `libflate`

jtgeibel opened the pull request:
Begin bisecting the segfault caused by a bulk `cargo update` by only
bumping a few dependencies.

`cargo update -p flate2:0.2.20 -p flate2:1.0.7 -p libflate -p tar --aggressive`

Mark-Simulacrum commented:
Can we not get the core file? Or get the trace from it somehow?

jtgeibel commented:
@Mark-Simulacrum I'm not aware of any easy way to get a core dump out of Heroku because the dyno is torn down and any temporary files deleted after the process exits.  It might be possible with a custom buildpack to detect a crash and upload the core file somewhere else, but I haven't really looked into that in depth.

Running the `all.rs` tests under `valgrind` flagged a few issues that were resolved by updating `flate2` in #1692.  I see no new issues flagged after running a full `cargo update`.

I've done some stress testing against the `server` binary and wasn't able to replicate the issue locally.  However, I was focused more on stressing overall concurrency at the time and was not hitting all of the endpoints (e.g., no endpoints that interacted with S3 or GitHub).  I had hoped that the issue had been resolved upstream (we've been slowly picking at this since #1263), but we had to revert #1756 after the issue popped back up in production.

Mark-Simulacrum commented:
Ah, okay. I was hoping heroku had some native tooling to deal with this. We can probably do it via a small wrapper binary written in Rust that spawns the real crates.io and waits for it to segfault. I'm not sure we can bump `ulimit -c` on Heroku though (or, really, how to do so from Rust).

If this doesn't work out I might be able to spend some time this weekend figuring that out for you, if you'd like -- just let me know.

sgrif commented:
I can ask if there's a way to get a core dump

smarnach commented:
Instead of writing a wrapper binary, it's probably easier to update the command in `Procfile` – the command is executed in bash, so we can add `ulimit -c` there, and some command to upload the dump somewhere – provided Heroku allows this in the first place.

smarnach commented:
For what it's worth, I poked around a bit on a free dyno, and it appears there is no way to generate a core dump, at least no obvious one.

* Both the hard and the soft limit for core dumps are set to zero.
* `/proc/sys/kernel/core_pattern` is set to its default value `core`, and we don't have permission to change it to pipe the core to a custom program.

I doubt that a custom buildpack will help, but it is possible that paid dynos have other limits than free ones.

jtgeibel commented:
I've added a commit to update `libflate`.  We pull in `libflate` (a Rust implementation) via `reqwest`.  `flate2` (C bindings) is used directly in the TOML file and via `conduit-git-http-backend`.  We may want to consolidate these in the future. 

I've updated the title and description to reflect that `libflate` is now updated as well.

bors commented:
:umbrella: The latest upstream changes (presumably #1783) made this pull request unmergeable. Please resolve the merge conflicts.

sgrif commented:
@bors: r+

bors commented:
:pushpin: Commit 95121b99c75a764d89f8616f85721836296ed7c1 has been approved by `sgrif`

<!-- @bors r=sgrif 95121b99c75a764d89f8616f85721836296ed7c1 -->
<!-- homu: {"type":"Approved","sha":"95121b99c75a764d89f8616f85721836296ed7c1","approver":"sgrif"} -->

bors commented:
:hourglass: Testing commit 95121b99c75a764d89f8616f85721836296ed7c1 with merge 75f99a060c125834c70ed9f627ab0d5a69070442...
<!-- homu: {"type":"BuildStarted","head_sha":"95121b99c75a764d89f8616f85721836296ed7c1","merge_sha":"75f99a060c125834c70ed9f627ab0d5a69070442"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/crates.io/builds/119540056)
Approved by: sgrif
Pushing 75f99a060c125834c70ed9f627ab0d5a69070442 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"sgrif","base_ref":"master","builders":{"checks-travis":"https://travis-ci.com/rust-lang/crates.io/builds/119540056"},"merge_sha":"75f99a060c125834c70ed9f627ab0d5a69070442"} -->
